### DISCUSSION TITLE
Click, CLI Commands, and entry_points

### DISCUSSION THREAD

**Discussion by gg2** [ckan/ckan#6379] [Q&A] [answered]
I'm trying to update an extension to work with CKAN 2.9. We had paster commands registered in the setup.py entry_points field. I have never worked with entry_points before. Our entry_points looks like this:

```
entry_points='''
    [ckan.plugins]
    template=ckanext.template.plugin:TemplatePlugin
    
    [paste.paster_command]
    template=ckanext.template.commands.custom_commands:CustomCommand
''',
```

* First off, what is this string syntax? I can't find any references to setup.py entry_points that uses this syntax.
* Secondly, more importantly, what should [paste.paster_command] be updated to? `[ckan.cli]`? Something to do with Click? Don't update it at all?
* Thirdly, since I've changed to Click, I modified the commands structure as best I could tell how. (I'm not able to test yet, because I don't have the plugin converted entirely yet. Though, I probably should set up a new template to test just this part by itself... There's a thought.) Anyway, the command was a class before: `CustomCommand`, with its main method splitting off into other methods based on the args it received. I'm revising it now with Click's groups and subcommands. The main group is: 

```
@click.group(name=u'custom_command', short_help=u'Custom commands to manage CKAN')
def custom_command():
    pass
```

So that means the command line in the first entry_points example above can become something like this?

```
    [ckan.cli?]
    custom_command=ckanext.template.commands.custom_commands:custom_command
```

To note: I am using `template=ckanext...` above and `custom_command=ckanext...` below on purpose in my examples.

**Comment by Zharktas** [comment, marked as answer]
The syntax is in .ini-style https://setuptools.readthedocs.io/en/latest/pkg_resources.html#creating-and-parsing  ```If data is a string or sequence of lines, it is first split into .ini-style sections (using the split_sections() utility function) and the section names are used as group names.```

Click commands implemented with IClick interface do not require entry in entry_points anymore as they are executed through the new ckan command instead of paster which required that you tell paster about them via entry_points. 

You can just remove whole entry if you want to support only CKAN 2.9.

**Comment by gg2** [comment]
Thank you, on all points! That makes sense, and I appreciate the reference to where in setuptools docs this is mentioned.
